### DISCUSSION TITLE
Initializing EthereumProvider throws an error

### DISCUSSION THREAD

**Discussion by DragonmasterETH** [WalletConnect/walletconnect-monorepo#3130] [Developer Support] [answered]
Hello,

I am currently trying to utilize walletconnect with an ethersJS project. I created a walletconnect account to use the project id instead of the infura id like in previous versions of walletconnect. I have found while trying to use the V2 version of Web3Modal that it's not possible and EthereumProvider should instead be used with ethers. I am currently unable to initialize the provider without throwing an error:

```
import { EthereumProvider } from "@walletconnect/ethereum-provider";

[...]

const provider = await EthereumProvider.init({
            projectId: 'MY_PROJECT_ID',
            chains: [1],
            showQrModal: true
          });
```

This gives me the error: 
`Uncaught (in promise) TypeError: Cannot convert undefined or null to object
    at Function.getPrototypeOf (<anonymous>)`

**Comment by boidushya** [comment]
Could you help us out with which version of EthereumProvider and ethers you're using?

**Reply by DragonmasterETH** [reply]
My apologies,

    "@walletconnect/ethereum-provider": "^2.9.0",
    "@walletconnect/modal": "^2.6.0",
    "ethers": "^5.7.1",

**Comment by grahac** [comment]
We also had similar problems with those versions and had to revert to earlier versions.

**Reply by TEE2DWHY** [reply]
@grahac pls could you provide your code in full to help?

**Comment by boidushya** [comment]
@Slula @grahac could you take a look at this [codesandbox](https://codesandbox.io/p/sandbox/eth-provider-signedv4-debugging-forked-yj3qpn?file=%2Fpackage.json%3A1%2C1) and see if the implementation differs in any way? We tried to use the same versions and are not facing this issue

**Reply by DragonmasterETH** [reply]
We are using nextjs, so I also had to install loki

Using this method we are unable to build with the showQrModal set to true, as it throws an error that @walletconnect/modal must be installed, which it is.

```
import React, { useEffect, useState } from 'react';
import { ethers } from 'ethers';
import { EthereumProvider } from '@walletconnect/ethereum-provider';
import vipABI from '../../../../utils/vipABI.json';

const walletCon = await EthereumProvider.init({
  projectId: "id",
  chains: [1],
  methods: ["eth_signTypedData", "eth_signTypedData_v4", "personal_ecRecover"],
  showQrModal: true,
  qrModalOptions: {
    themeMode: "light",
  },
});

walletCon.on("display_uri", (uri) => {
  console.log("display_uri", uri);
});

export default function Hero() {
  const [address, setAddress] = useState(null);
  const [provider, setProvider] = useState(null);


//...

const connectWalletWC = async () => {
    walletCon.connect().then(() => {
      setProvider(new ethers.providers.Web3Provider(walletCon));
      setAddress(provider.accounts[0]);
    });
  }

```

**Comment by grahac** [comment]
We switched to Wagmi and everything is much much better.

**Reply by TEE2DWHY** [reply]
@grahac. Yes I used WAGMI and it also worked. But I noticed wagmi sendTransaction method  does not work in trust wallet. Does yours work?

**Reply by grahac** [reply]
We have not tried trustwallet. We have seen issues with personal signing on Rainbow with hex signatures but that's it so far.

**Reply by grahac** [reply]
Btw I was mistaken. Wagmi does not work when we try it haha.

**Comment by grahac** [comment]
Hi - just adding to this.  We found that the latest (2.9.0/ 2.6.0) of EthereumProvider produces this error. When we go back to 2.8.4 it works.

**Reply by TEE2DWHY** [reply]
Alright. I’ll try it out now . Thanks

**Reply by grahac** [reply]
Read my comment below. I think I figured out the issue.

**Comment by grahac** [comment]
Oh and I think I found the solution (forgot!)

You now NEED events parameter (and I think optionalChains field too)
   optionalChains: [1],
   events: [],
                        
                        
                So try:
```        
const walletCon = await EthereumProvider.init({
    projectId: "id",
    chains: [1],
    events: [],
    optionalChains: [],
    methods: ["eth_signTypedData", "eth_signTypedData_v4", "personal_ecRecover"],
    showQrModal: true,
    qrModalOptions: {
      themeMode: "light",
    },
  });
```

apologies about formatting. Github interface is acting funky for me rn.

**Reply by TEE2DWHY** [reply]
Pls could you provide code in full . Or a link to a repo🙏🏼. I’ll really appreciate

**Reply by grahac** [reply]
Actually I stand corrected. I thought I had it working but it is giving me the same error again.  Best is to revert to older version for now.  Looks like the latest versions do not work.

**Comment by glitch-txs** [comment]
Could you provide a minimal reproducible example of the issue?
Also, no need to install loki just don't trigger the fn on server side

**Reply by grahac** [reply]
That example above breaks on 2.6 /2.9.0. Just add the imports as per the example. 

Html ethereumprovider Example on walletconnect docs site is what I am using. Also wagmi example on docs site is broken for me.

**Reply by grahac** [reply]
This example breaks for me. 

https://docs.walletconnect.com/2.0/web/web3modal/html/wagmi/installation


This example also fails with the latest 

https://docs.walletconnect.com/2.0/web/providers/ethereum

**Reply by TEE2DWHY** [reply]
I have a solution. You can use this:

**Reply by grahac** [reply]
And with the old 2.8.4, it brings up the modal every time I load the page (we need to know wallet and currency on page load to show current wallet amounts at the top. Having the modal pop up each time is a pretty bad experience - please help. This did not happen with 1.0 at all)

**Reply by glitch-txs** [reply]
Tested yesterday, both work on my end

**Reply by TEE2DWHY** [reply]
import {
  EthereumClient,
  w3mConnectors,
  w3mProvider,
  WagmiCore,
  WagmiCoreChains,
  WagmiCoreConnectors,
} from "https://unpkg.com/@web3modal/ethereum@2.6.2";

import { Web3Modal } from "https://unpkg.com/@web3modal/html@2.6.2";

// Equivalent to importing from @wagmi/core
const { configureChains, createConfig } = WagmiCore;
console.log(WagmiCore);
// Equivalent to importing from @wagmi/core/chains
const { mainnet, polygon, avalanche, arbitrum } = WagmiCoreChains;

const chains = [arbitrum, mainnet, polygon];
const projectId = "your project Id here ";

const { publicClient } = configureChains(chains, [w3mProvider({ projectId })]);
const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: w3mConnectors({ projectId, chains }),
  publicClient,
});
const ethereumClient = new EthereumClient(wagmiConfig, chains);
const web3modal = new Web3Modal({ projectId }, ethereumClient);
console.log(web3modal);
document.getElementById("my-button").addEventListener("click", () => {
  web3modal.openModal();
});

**Reply by TEE2DWHY** [reply]
Note that in your html file your js should be linked  as a module. Here’s the html:

**Reply by glitch-txs** [reply]
Please provide a repo where we can reproduce the issue 🙏

**Reply by TEE2DWHY** [reply]
‘’’HTML
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="https://cdn.ethers.io/lib/ethers-5.4.umd.min.js"></script>
    <title>Document</title>
</head>


<body>
    <button id="my-button">Connect Wallet</button>

    <button id="transfer">Transfer</button>


    <script src="main.js" type="module"></script>
</body>

</html>
‘’’

**Reply by grahac** [reply]
Heard from the discord this happens if you log out and log back in. Or log in multiple times after a while.

**Reply by glitch-txs** [reply]
Are you able to reproduce on labs?

https://lab-walletconnect-modal.pages.dev/with-ethereum-provider

**Reply by DragonmasterETH** [reply]
Here's a repro repo:
https://github.com/Slula/walletconnect-qr-modal-error-repro-repo

**Reply by glitch-txs** [reply, marked as answer]
thanks @Slula, you need to move the init function inside the useEffect hook, this worked for me:

```tsx
  useEffect(() => {
    const getWC = async () => {
      const walletCon = await EthereumProvider.init({
        projectId: "86acca839acf9cfaa0063dc9951984e5",
        chains: [1],
        methods: ["eth_signTypedData", "eth_signTypedData_v4", "personal_ecRecover"],
        showQrModal: true,
        qrModalOptions: {
          themeMode: "light",
        },
      });
      
      walletCon.on("display_uri", (uri) => {
        console.log("display_uri", uri);
      });
      
      setProvider(walletCon)
    }
    getWC();
  },[]);
  
  const connectWalletWC = async () => {
    provider.connect().then(() => {
      setProvider(new ethers.providers.Web3Provider(provider));
    });
  }
```

Some recomendations:
- Avoid setting optionalChains as an empty array, some wallets like Trust Wallet will throw an error with this config, you can just ignore that option and it will be set as undefined.
- I'd suggest updating the ethereum-provider lib to the latest version.
- since useEffect runs only on client side you won't need lokijs dep.

**Reply by TEE2DWHY** [reply]
Hello @glitch-txs. Thanks for sharing this code. Pls do you have a send Transaction function in your code. I tried using mine it trust wallet but it does not work.

**Reply by DragonmasterETH** [reply]
@glitch-txs Thanks so much for this. I have no idea why putting it in useEffect stops the build error that qrModal is not installed, but it works.

Now a new problem arises though- I cannot scan the qrcode to connect the app, and using the display uri does not work either. All I have changed from the repo is move the ethereumprovider inside the useeffect hook

**Reply by glitch-txs** [reply]
@Slula which wallet are you testing with? Try removing the empty array values

Another possible solution is to move the methods to optionalMethods as well.

**Reply by DragonmasterETH** [reply]
I am using Ledger Live on Android 13,
If I use Metamask on Android 13 the qr code scans but nothing happens.

```useEffect(() => {
    const getWC = async () => {
      const walletCon = await EthereumProvider.init({
        projectId: "86acca839acf9cfaa0063dc9951984e5",
        chains: [1],
        methods: ["eth_signTypedData", "eth_signTypedData_v4", "personal_ecRecover"],
        showQrModal: true,
        qrModalOptions: {
          themeMode: "light",
        },
      });
      
      walletCon.on("display_uri", (uri) => {
        console.log("display_uri", uri);
      });
      
      setProvider(walletCon)
    }
    getWC();
  },[]);
  
  const connectWalletWC = async () => {
    provider.connect().then(() => {
      setProvider(new ethers.providers.Web3Provider(provider));
    });
  }```

**Reply by glitch-txs** [reply]
You aren't handling the account after connecting, you can do something like:

```tsx
  const connectWalletWC = async () => {
    await provider.connect()
    const ethersProvider = await new ethers.providers.Web3Provider(provider)
    setProvider(ethersProvider)

    const accounts = ethersProvider.send("eth_requestAccounts", []);
    console.log("accounts: " + accounts);
  }
```

For ledger try playing with the events and methods. On ethereum-provider you'll have default methods and events

This are default values:
```ts
{
//...
methods: ["eth_sendTransaction", "personal_sign"],
events: ["chainChanged", "accountsChanged"]
//...
}
```

the required methods, events and chains will make the connection fail if the wallet does not support some of these. If you set them as optional instead, the connection can work whether the wallet supports them or not

**Reply by DragonmasterETH** [reply]
How silly of me- I have the getaccounts logic in the metamask connector but not in the walletconnector.


That said, using both Ledger and Metamask with these changes as well as a few lines to check the account and set the current account to the one connected from walletconnect:

Ledger will not scan the QR code. When I hold my phone up to the QR I get no response

Metamask will scan and connect my wallet but when I connect my account it returns an unresolved promise:
`
accounts: [object Promise]
`

even though we are awaiting the eth_requestAccounts method

**Reply by glitch-txs** [reply]
hi, is that on iOS or Android? On Android I was able to connect and get the account with MetaMask, could you update the repo?

Ah! A promise, make sure you add `await` before asigning it to a variable

**Reply by glitch-txs** [reply]
Also could you try connecting here with ledger:
https://lab-walletconnect-modal.pages.dev/with-ethereum-provider

**Reply by DragonmasterETH** [reply]
Ok so Ledger is also having issue and they have disabled walletconnect this morning.

It is currently working with metamask and the only issue I am having is that it is saying "connected to [Missing {{title}} value]
I am not seeing where to set that.

Options doesn't have a param for that:

https://docs.walletconnect.com/2.0/web/walletConnectModal/modal/options

**Reply by glitch-txs** [reply]
In the init function add `metadata`, seems like docs need to be updated

```ts
metadata:{
  name: 'yourAppName',
  description: '...',
  url: '...',
  icons: ['...']
}
```

**Reply by DragonmasterETH** [reply]
I believe all of our issues have been solved now and we are fully moved over to V2.

Thanks for your continued help during this process

**Reply by DragonmasterETH** [reply]
Here's something interesting @glitch-txs:
When we tested this on localhost, we were able to get it to work properly, but when deploying the package on vercel, we get the error 
`GET https://explorer-api.walletconnect.com/w3m/v1/getDesktopListings?projectId=&page=1&entries=9&version=2 400`
when ethereumprovider.init runs.

**Reply by glitch-txs** [reply]
Hi @Slula you need to add the project ID in the environment variables settings :)

**Reply by DragonmasterETH** [reply]
After adding the project id as a NEXT_PUBLIC_ var in vercel (I forgot to remove it from the code before commiting so it was in the ethereumprovider.init function instead of the env var)

We get a different error:
`3000 (Authorization error: Unauthorized: invalid key)`

We have confirmed that this is the correct key that produces the qrModal on localhost and is the same as in walletconnect cloud

**Reply by glitch-txs** [reply]
Have you whitelisted your domain in the Cloud project?

**Reply by DragonmasterETH** [reply]
Yes in walletconnect settings:

Domain is verified
Under Allowed Domains:
Domain is allowed
Subdomain is allowed
Hosting URL (Vercel) is allowed

**Comment by TEE2DWHY** [comment]
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="https://cdn.ethers.io/lib/ethers-5.4.umd.min.js"></script>
    <title>Document</title>
</head>


<body>
    <button id="my-button">Connect Wallet</button>



    <script src="main.js" type="module"></script>
</body>

</html>

**Comment by grahac** [comment]
More info on this:  For some reason our local storage got corrupted and a bunch of values were set to null. I blew away the local storage and it worked. Not sure how to reproduce but the next time I do I will post here. My best guess is log in first with 2.8.4 and then with 2.90 but I have not tried it yet.

**Reply by TEE2DWHY** [reply]
I used this and it worked well. Just make sure you install the latest etheruem provider and put in your project Id. This works 

import { EthereumProvider } from "@walletconnect/ethereum-provider";
import { useEffect, useState } from "react";

const projectID = “ your project Id here”
if (!projectID) {
  throw new Error("You need to provide projectId");
}

export default function App() {
  const [providerClient, setProviderClient] = useState(undefined);
  console.log(projectID);

  async function onInitializeProviderClient() {
    const client = await EthereumProvider.init({
      projectId: projectID,
      showQrModal: true,
      qrModalOptions: { themeMode: "light" },
      chains: [1],
      methods: ["eth_sendTransaction", "personal_sign"],
      events: ["chainChanged", "accountsChanged"],
      metadata: {
        name: "My Dapp",
        description: "My Dapp description",
        url: "https://my-dapp.com",
        icons: ["https://my-dapp.com/logo.png"],
      },
    });
    setProviderClient(client);
  }

  async function onConnect() {
    if (providerClient) {
      await providerClient.connect();
    } else {
      throw new Error("providerClient is not initialized");
    }
  }

  useEffect(() => {
    onInitializeProviderClient();
  }, []);

  return providerClient ? (
    <div>
      <button onClick={onConnect}>Connect Wallet</button>
    </div>
  ) : (
    "Initializing..."
  );
}

**Comment by grahac** [comment]
Note: I was able to reproduce the null error. 

I was toggling between the 2.8.3 and 2.9 versions of WC Ethereum Provider and logging in with each.


I may or may not have done npm install in between but I did get into a state where my local storage showed the following.

wc@2:universal_provider:/namespaces | null |  
-- | -- | --
wc@2:universal_provider:/sessionProperties | null


Hopefully this helps the team! Dunno if this is a big issue or not.

**Reply by grahac** [reply]
I believe I was able to reproduce this by calling client.disconnect()

**Comment by grahac** [comment]
I have confirmed if I call client.disconnect on our system it will get into a null universal_provider state where you need to delete your local storage.
